### PR TITLE
gh-91351: Fix some bugs in importlib handling of re-entrant imports

### LINKED ISSUE #91351: importlib lock race issue in deadlock handling code
BPO | [47195](https://bugs.python.org/issue47195)
--- | :---
Nosy | @rpurdie
Files | <li>[testit2.py](https://bugs.python.org/file50714/testit2.py "Uploaded as text/plain at 2022-04-01.20:23:23 by @rpurdie")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2022-04-01.20:23:24.167>
labels = ['interpreter-core', '3.10']
title = 'importlib lock race issue in deadlock handling code'
updated_at = <Date 2022-04-02.06:56:46.108>
user = 'https://github.com/rpurdie'
```

bugs.python.org fields:
```python
activity = <Date 2022-04-02.06:56:46.108>
actor = 'rpurdie'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core']
creation = <Date 2022-04-01.20:23:24.167>
creator = 'rpurdie'
dependencies = []
files = ['50714']
hgrepos = []
issue_num = 47195
keywords = []
message_count = 2.0
messages = ['416517', '416539']
nosy_count = 1.0
nosy_names = ['rpurdie']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue47195'
versions = ['Python 3.10']
```

</p></details>

### Linked PRs
* gh-94504

### PR DESCRIPTION
**Pull request by exarkun**
Fixes #91351 
Update of #94342 

See https://github.com/python/cpython/issues/91351 for details about the problem.

Re-entrancy is always tricky and given the requirements of _bootstrap.py (to operate with re-entrancy and multi-threading and to do so without exposing any of the details to application code doing an import) I think this goes double.

This PR does a few things to achieve _better_ safety in the face of re-entrancy:

* Switch some data structures to those that support atomic operation so that they are consistent in case of asynchronous re-entrancy (eg from the garbage collector or a signal handler).
* Use RLock and add more RLock-like behavior to prevent deadlocks in the re-entrant case.
* Update the deadlock detection algorithm to support the fact that one thread might be "blocked on" acquiring the module lock for more than one module at a time.

I'm not quite sure I believe this new version of the code is 100% correct with respect to re-entrancy but it does fixes mishandling of two specific cases:

* A re-entrant import is performed between the time `_blocking_on` is populated and cleaned up inside `_ModuleLock.acquire`.  Previous this failed with a `KeyError` (as described in the linked issue).
* A re-entrant import is performed while the module lock is held inside `_ModuleLock.acquire`.  Previously this failed by deadlocking.

This PR also does not include any new unit tests.  I have a small stand-alone program which can reproduce both of these but only with the assistance of some additional instrumentation inside `_bootstrap.py` to make sure the re-entrancy happens at the interesting times.  If adding this kind of instrumentation is acceptable then it may be possible to turn this program into some unit tests.

Compared to the previous PR, this one is a bit simpler because it uses `RLock` in a place where main currently uses a regular `Lock`.

For reference, here is a stand-alone reproducer.  This one very reliably reproduces the `KeyError` problem (by exercising the codepath forever until it hits it).  I haven't managed to create a deterministic reproducer for the deadlock case - except with the assistance of instrumentation inside `_bootstrap.py` itself.

```
import sys, socket, gc

class Cycle:
    pass

def a_cycle():
    c = Cycle()
    c.cycle = c
    c.s = socket.socket()

def main():
    while True:
        # import a module that socket.__del__ is going to import to exercise
        # re-entrant _ModuleLock.lock handling
        a_cycle()
        import linecache

        del sys.modules["linecache"]

main()
```

* Issue: gh-91351

### CODE REVIEW

**Comment by rpurdie** [comment]
For what it is worth, Yocto Project keep running into issues caused by re-entrant imports as mentioned in issue #91351. I believe that issue references a couple of older issues where people have run into this too. Is it possible to review this and get it resolved?

**Comment by brettcannon** [comment]
@rpurdie I do plan to review it, but it is not at the top of [my review queue](https://github.com/pulls/review-requested) yet. If you want to provide a review yourself then that would be appreciated.

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
    """A context manager responsible to updating ``_blocking_on``."""
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
    def __init__(self, thread_id, lock):
        self.thread_id = thread_id
```

I realize there are other places that use `tid` for a variable name, but I don't think that's descriptive enough on its own.

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
```

I think the comment is redundant as what `_blocking_on` is used for is explained nicely above and the variable name is self-explanatory.

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        """Mark the running thread as waiting for self.lock. via _blocking_on."""
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # re-entrant (i.e., a single thread may take it more than once) so it
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        self.blocked_on = _blocking_on.setdefault(self.tid, [])
        self.blocked_on.append(self.lock)
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
    def __exit__(self, *args, **kwargs):
        """Remove self.lock from this thread's _blocking_on list."""
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
def _is_deadlocked(seen, subject, thread_ids, _blocking_on):
```

`_has_deadlock` reads to me like a deadlock is a noun and something you can hold in your hand. But to me, deadlock is a state, and thus something you can be.

And what is `subject`? and `seen` what? I think more descriptive parameters would be helpful.

And why is `_blocking_on` a parameter? And since it is a parameter, there's no need for the leading `_`.

I would also change the order to put `subject` first since that's the key object for this method.

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]: The stdlib doesn't mark parameters in docstrings like this. See https://peps.python.org/pep-0257/ for how to write docstrings that meet our style guide.

BTW, this is also why having a newline after `"""` is being changed by me in a lot of comments.

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
    """Check if 'subject' is holding the same lock as another thread(s).
    
    The search within _blocking_on starts with the threads listed in tids.
    'seen' contains any threads that are considered already traversed in the search.
    
    """
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        if not blocking_on:
```

No need to be specific about `None` or `[]` in this case; same outcome.

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
            # eventually leads to a fixed point or a cycle, but
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]: What is `me`? Is that meant to be what `subject` represents?

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]: Is it worth doing this recursively? From the one use of the function below it looks like reasonable defaults could be provided and then you would use a `while` loop until the candidate threads to check is exhausted. You could then have an `else` clause that returns `False` as the default result.

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # corresponding module.  Since it is an RLock, a single thread will be
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # Represent the number of times the owning thread has acquired this lock
        # via a list of `True`.  This supports RLock-like ("re-entrant lock")
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]: I would consider using `True` so that `sum()` also works in this instance.

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # and list.pop() are both atomic and thread-safe in CPython and it's hard
        # to find another primitive with the same properties.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
                        raise _DeadlockError(f'deadlock detected by {self!r}')
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
                    # It's not clear who is expected to handle this error.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
                # Taking the lock has served its purpose (making us wait), so we can
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
                # next iteration around this 'while' loop.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
            if not len(self.count):
```

**Review by brettcannon — CHANGES_REQUESTED** [review]

File: Lib/importlib/_bootstrap.py
Comment (exarkun) [reply Lib/importlib/_bootstrap.py]: I agree it could be implemented iteratively.  This seems like a purely stylistic choice though.  Unless there is a stronger motivation I would prefer not to rewrite it.

**Comment by exarkun** [comment]
I have made the requested changes; please review again

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
# A dict mapping thread IDs to lists of _ModuleLock instances.  This maps a
# thread to the module locks it is blocking on acquiring.  The values are
# lists because a single thread could perform a re-entrant import and be "in
# the process" of blocking on locks for more than one module.  A thread can
# be "in the process" because a thread cannot actually block on acquiring
# more than one lock but it can have set up bookkeeping that reflects that
# it intends to block on acquiring more than one lock.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
            continue
        elif tid in seen_ids:
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # If a different thread than the running one holds the lock then the
        # thread will have to block on taking the lock, which is what we want
        # for thread safety.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # Can end up being set to None if this lock is not owned by any thread
        # or the thread identifier for the owning thread.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # `self.wakeup.acquire()`, awaiting to get their turn holding this module
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # look at _blocking_on to see if any threads are blocking
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
            # Try to find this thread.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
            # Start from the thread that holds the import lock for this
            # module.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
            # Use the global "blocking on" state.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
                # Now take the lock in a blocking fashion.  This won't
                # complete until the thread holding this lock
                # (self.owner) calls self.release.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
            # bpo 38091: the chain of tid's we encounter here eventually leads
            # to a fixed point or a cycle, but does not reach target_id.
            # This means we would not actually deadlock.  This can happen if
            # other threads are at the beginning of acquire() below.
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # self.wakeup.acquire() to try to get their turn holding this module
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # via a list of True.  This supports RLock-like ("re-entrant lock")
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        # This is incremented in self.acquire() when a thread notices it is
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
                # Taking the lock has served its purpose (making us wait), so we can
                # give it up now.  We'll take it w/o blocking again on the
                # next iteration around this 'while' loop.
```

**Review by brettcannon — APPROVED** [review]

**Review by brettcannon — CHANGES_REQUESTED** [review]
Unfortunately it looks like I can't commit the comment changes myself (the box to allow maintainers to make changes to the branch not checked?), so at this point is just touching up the comments and then I think we can merge this!

**Comment by brettcannon** [comment]
@exarkun there's also a news entry missing from this PR which we will need before we can merge it.

**Comment by exarkun** [comment]
I have made the requested changes; please review again

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]: Because there are so many arguments and knowing the order to specify isn't necessarily intuitive, we should probably make this keyboard-only to force callers to spell out their arguments.

```suggestion
def _has_deadlocked(target_id, *, seen_ids, candidate_ids, blocking_on):
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        if not (candidate_blocking_on := blocking_on.get(tid)):
```

File: Lib/importlib/_bootstrap.py
Comment (brettcannon) [inline Lib/importlib/_bootstrap.py]:
```suggestion
        if _has_deadlocked(target_id, seen_ids=seen_ids, candidate_ids=edges,
                blocking_on=blocking_on):
```

**Review by brettcannon — APPROVED** [review]

**Review by brettcannon — CHANGES_REQUESTED** [review]
I would have approved and made the commits myself, but the branch is still blocked from me making changes. 😅 Can you just make the last couple and then I can merge this!

**Comment by exarkun** [comment]
I have made the requested changes; please review again

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
Thanks so much for the PR and sticking with it until it got in!

**Pull request merged**
